# Functions

Use `def` to "define" the name and variables of the function. Of note, your function does *not* need to have any parameters/arguments [See "Hello, world!" example below].

Use `return` to give the resulting output

Let's use the example of finding the area of a triangle.

Recall, $\text{triangle area} = 1/2 \times \text{base} \times \text{height}$

In [4]:
def triangle_area(base, height):
    area = (1 / 2) * base * height
    return area

No code is executed; it will only execute when the function is *called*.

In [5]:
a1 = triangle_area(3, 8)
print(a1)

a2 = triangle_area(14, 2)
print(a2)

12.0
14.0


Example:

Create a function that converts Fahrenheit to Celsius

recall, $\text{Celsius} = 5/9 \times (\text{Fahrenheit} - 32)$

In [6]:
def fahrenheit2celsius(fahrenheit):
    celsius = (5/9) * (fahrenheit - 32)
    return celsius

In [8]:
# Test

c1 = fahrenheit2celsius(32)

c2 = fahrenheit2celsius(212)

print(c1, c2)

0.0 100.0


You can also call other functions when creating a *new* function

Example: convert Fahrenheit to Kelvin knowing that $\text{Kelvin} = \text{Celsius} + 273.15$

In [9]:
def fahrenheit2kelvin(fahrenheit):
    celsius = fahrenheit2celsius(fahrenheit)
    kelvin = celsius + 273.15
    return kelvin

In [10]:
# Test

k1 = fahrenheit2kelvin(32)
k2 = fahrenheit2kelvin(212)

print(k1, k2)

273.15 373.15


## Argument-less Function

Example, a function that simply prints "Hello, world!" when called.

In [13]:
def hello():
    print("Hello, world!")

In [14]:
# Test

hello()

Hello, world!


# Visualizing Functions

The class provides a visualization mode using the CodeSkulptor website, which is useful for debugging and seeing what is going on under the hood. It is similar to PythonTutor. If you don't want to use CodeSkulptor (like me) consider using PythonTutor.com.

# Additional Operations

## Modular arithmetic

Example: Calculate the "tens" digit and the "ones" digit of a 2 digit number.

Hint: The modulo operator (`%`) gives the **remainder** following division

In [1]:
num = 49

tens = num // 10
ones = num % 10
print(tens, ones)

4 9


Example: Use `%` to for 24 hour clock questions

In [2]:
# what time will it be at the end of an 8 hour shift that begins at 2000?

hour = 20
shift = 8

print((hour + shift) % 24)

4




# Logic and Conditionals

## Boolean Logic
- `not`
- `and`
- `or`

In [6]:
a = True
b = False

print(not a)
print(a and b)
print(a or b)

False
False
True


## Comparison Operators

>
<
>=
<=
==
!=

In [7]:
a = 7 > 3
print(a)

True


In [8]:
x = 8
y = 5
b = x >= y

print(b)

True


In [9]:
# works with strings too
c = "Hello" == "Hello"

print(c)

True




## Conditionals

In [11]:
# If person is a friend (friend = TRUE) and you have > 20 bucks, then say Hi and give 20 bucks


def greet(friend, money):
    if friend and (money > 20):
        print("Hi!")
        money = money - 20
    elif friend:
        print("Hello")
    else:
        print("Ha ha")
        money = money + 10 #mug non-friend and take 10 bucks
    return money

In [13]:
money = 15

money = greet(True, money)
print("Money:", money)
print("") # adds a space between each line

money = greet(False, money)
print("Money:", money)
print("")

money = greet(True, money)
print("Money:", money)
print("")

Hello
Money: 15

Ha ha
Money: 25

Hi!
Money: 5



### Leap Year Example

In [14]:
def is_leap_year(year):
    if (year % 400) == 0:
        return True
    elif (year % 100) == 0:
        return False
    elif (year % 4) == 0:
        return True
    else:
        return False

In [15]:
year = 2012
leap_year = is_leap_year(year)
    
if leap_year:
    print(year, "is a leap year")
else:
    print(year, "is not a leap year")

2012 is a leap year


In [45]:
# Quiz Prob 8

import math

def polygon(n, s):
    ans = (n * s**2) / (4 * math.tan((math.pi/n)))
    return ans

In [47]:
polygon(7, 3)

32.705211996014306

In [58]:
def project_to_distance(point_x, point_y, distance):
    dist_to_origin = math.sqrt(point_x ** 2 + point_y ** 2)
    scale = distance / dist_to_origin
    print(point_x * scale, point_y * scale)

project_to_distance(2, 7, 4)

1.0988845115895123 3.846095790563293


Quiz 1

1. Unlimited
2. False False True False
3. Hello world     Is it over yet?
4. [2 choices] (n % 100 - n % 10) / 10     and      ((n - n % 10) % 100) / 10
5. 10
6. 69
7. 1061.8348011259045 
8. 32.705211996014306
9. Incorrect indentation
10. 3.846095790563293